apps/omg_watcher_info/lib/omg_watcher_info/utxo_selection.ex

```diff
@@ -17,95 +17,98 @@ defmodule OMG.WatcherInfo.UtxoSelection do
-  # We return {token, {change, [utxos for payment]}}, change > 0 means insufficient funds.
-  # NOTE: order of Utxo list is implicitly assumed for the algorithm to work deterministically,
-  # see: `OMG.WatcherInfo.DB.TxOutput.get_sorted_grouped_utxos`
-  @spec select_utxo(%{Transaction.Payment.currency() => list(%DB.TxOutput{})}, %{
-          Transaction.Payment.currency() => pos_integer()
-        }) ::
-          list({Transaction.Payment.currency(), {integer, list(%DB.TxOutput{})}})
-  defp select_utxo(utxos, needed_funds) do
+  defp prioritize_utxos_by_currency({currency, _utxos}, utxos, selected_utxo_hashes) do
+    utxos[currency]
+    |> filter_unselected(selected_utxo_hashes)
+    |> Enum.sort_by(fn utxo -> utxo.amount end, :asc)
+  end
+
+  @doc """
+  Given a map of UTXOs sufficient for the transaction and a set of available UTXOs,
+  adds UTXOs to the transaction for "stealth merge" until the limit is reached or
+  no UTXOs are available. Agnostic to the priority ordering of available UTXOs.
+  Returns an updated map of UTXOs for the transaction.
+  """
+  @spec add_utxos_for_stealth_merge(utxo_list_t(), utxos_map_t()) :: utxos_map_t()
+  def add_utxos_for_stealth_merge([], selected_utxos), do: selected_utxos
+
+  def add_utxos_for_stealth_merge(available_utxos, selected_utxos) do
+    case get_number_of_utxos(selected_utxos) do
+      Transaction.Payment.max_inputs() ->
+        selected_utxos
+
+      _ ->
+        [priority_utxo | remaining_available_utxos] = available_utxos
+
+        stealth_merge_utxos =
+          Map.update!(selected_utxos, priority_utxo.currency, fn current_utxos ->
+            [priority_utxo | current_utxos]
+          end)
+
+        add_utxos_for_stealth_merge(remaining_available_utxos, stealth_merge_utxos)
+    end
+  end
+
+  @doc """
+  Given the available set of UTXOs and the needed amount by currency, tries to find a UTXO that satisfies the payment with no change.
+  If this fails, starts to collect UTXOs (starting from the largest amount) until the payment is covered.
+  Returns {currency, { variance, [utxos] }}. A `variance` greater than zero means insufficient funds.
+  The ordering of UTXOs in descending order of amount is implicitly assumed for this algorithm to work deterministically.
+  """
+  @spec select_utxo(%{currency_t() => pos_integer()}, utxos_map_t()) ::
+          list({currency_t(), {integer, utxo_list_t()}})
+  def select_utxo(needed_funds, utxos) do
     Enum.map(needed_funds, fn {token, need} ->
       token_utxos = Map.get(utxos, token, [])
 
       {token,
        case Enum.find(token_utxos, fn %DB.TxOutput{amount: amount} -> amount == need end) do
          nil ->
            Enum.reduce_while(token_utxos, {need, []}, fn
-             _, {need, acc} when need <= 0 -> {:halt, {need, acc}}
-             %DB.TxOutput{amount: amount} = utxo, {need, acc} -> {:cont, {need - amount, [utxo | acc]}}
+             _, {need, acc} when need <= 0 ->
```

Comment:
This is a reduce_while in a case in a map, I know it's not from you but if you can simply it a bit (split in functions?) it'd be great